In [6]:
import sys
from pathlib import Path

# Notebook cell
%load_ext autoreload
%autoreload 2

# Get root directory (assuming notebook is in root/notebooks/)
NOTEBOOK_DIR = Path.cwd()
ROOT_DIR = NOTEBOOK_DIR.parent if NOTEBOOK_DIR.name == 'notebooks' else NOTEBOOK_DIR

# Add src directory to Python path
sys.path.append(str(ROOT_DIR / 'src'))

# Verify path
print(f"Python will look in these locations:\n{sys.path}")


# --- Execute the processor ---
import utils

# SOURCE_PATH_OHLCV = '..\data\df_OHLCV_2025-03-07_clean.pkl'
# SOURCE_PATH_STOCK = '..\data\df_finviz_stocks_n_ratios.pkl'
# SOURCE_PATH_ETF = '..\data\df_finviz_etfs_n_ratios.pkl'

# SOURCE_PATH, DEST_PATH = utils.main_processor(
#     data_dir='..\data',  # search project ..\data
#     downloads_dir='',  # None searchs Downloads dir, '' omits search
#     downloads_limit=0,  # search the first 10 files
#     clean_name_override='df_perf_ratios.pkl',  # override filename
#     start_file_pattern='df_', # search for files starting with 'df_'
# )



Python will look in these locations:
['C:\\Users\\ping\\.pyenv\\pyenv-win\\versions\\3.10.5\\python310.zip', 'C:\\Users\\ping\\.pyenv\\pyenv-win\\versions\\3.10.5\\DLLs', 'C:\\Users\\ping\\.pyenv\\pyenv-win\\versions\\3.10.5\\lib', 'C:\\Users\\ping\\.pyenv\\pyenv-win\\versions\\3.10.5', 'c:\\Users\\ping\\Files_win10\\python\\py310\\.venv', '', 'c:\\Users\\ping\\Files_win10\\python\\py310\\.venv\\lib\\site-packages', 'c:\\Users\\ping\\Files_win10\\python\\py310\\.venv\\lib\\site-packages\\win32', 'c:\\Users\\ping\\Files_win10\\python\\py310\\.venv\\lib\\site-packages\\win32\\lib', 'c:\\Users\\ping\\Files_win10\\python\\py310\\.venv\\lib\\site-packages\\Pythonwin', 'c:\\Users\\ping\\Files_win10\\python\\py310\\stocks\\src']


In [7]:
import pandas as pd

_df = pd.read_pickle('..\data\df_finviz_stocks_n_ratios.pkl')
df_stocks = _df.iloc[0:100].copy()
display(df_stocks.info())

# Convert DataFrame to list of dictionaries with index as 'Ticker'
records_stocks = df_stocks.reset_index().to_dict('records')

# Convert to string representation
records_stocks_str = str(records_stocks)

# Print first 1000 characters to preview
display(records_stocks_str[:1000])

<class 'pandas.core.frame.DataFrame'>
Index: 100 entries, AAPL to VRTX
Data columns (total 40 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Industry       100 non-null    string 
 1   Market Cap     100 non-null    float64
 2   P/E            99 non-null     float64
 3   Fwd P/E        100 non-null    float64
 4   PEG            94 non-null     float64
 5   P/FCF          92 non-null     float64
 6   Dividend       86 non-null     float64
 7   Payout Ratio   100 non-null    float64
 8   EPS this Y     100 non-null    float64
 9   EPS next Y     100 non-null    float64
 10  EPS next 5Y    99 non-null     float64
 11  Sales past 5Y  100 non-null    float64
 12  Sales Q/Q      99 non-null     float64
 13  EPS Q/Q        99 non-null     float64
 14  Inst Own       100 non-null    float64
 15  ROA            100 non-null    float64
 16  ROE            97 non-null     float64
 17  Curr R         96 non-null     float64
 18  Quick R    

None

"[{'Ticker': 'AAPL', 'Industry': 'Consumer Electronics', 'Market Cap': 3535.14, 'P/E': 37.41, 'Fwd P/E': 28.72, 'PEG': 3.8, 'P/FCF': 35.96, 'Dividend': 0.0044, 'Payout Ratio': 0.1611, 'EPS this Y': 0.0845, 'EPS next Y': 0.1192, 'EPS next 5Y': 0.0984, 'Sales past 5Y': 0.0918, 'Sales Q/Q': 0.0395, 'EPS Q/Q': 0.1013, 'Inst Own': 0.6377, 'ROA': 0.2757, 'ROE': 1.3652, 'Curr R': 0.92, 'Quick R': 0.88, 'Gross M': 0.4652, 'Oper M': 0.3176, 'Profit M': 0.243, 'Beta': 1.25, 'ATR': 5.98, 'Volatility W': 0.032, 'Volatility M': 0.0229, 'Recom': 2.14, 'Target Price': 254.02, 'Sharpe 30d': 2.3426617357009416, 'Sortino 30d': 3.7633252796759735, 'Omega 30d': 1.4917237958637106, 'Sharpe 63d': -0.293603939661817, 'Sortino 63d': -0.39393272544815183, 'Omega 63d': 0.951220126611859, 'Sharpe 125d': 0.6798203966962234, 'Sortino 125d': 0.9494529310191717, 'Omega 125d': 1.1233388717662842, 'Sharpe 250d': 1.3980668604113802, 'Sortino 250d': 2.120120547140081, 'Omega 250d': 1.2751593411559838}, {'Ticker': 'MSFT'

In [8]:
import pandas as pd

_df = pd.read_pickle('..\data\df_finviz_etfs_n_ratios.pkl')
df_etfs = _df.iloc[0:40].copy()
display(df_etfs.info())

# Convert DataFrame to list of dictionaries with index as 'Ticker'
records_etfs = df_etfs.reset_index().to_dict('records')

# Convert to string representation
records_etfs_str = str(records_etfs)

# Print first 1000 characters to preview
display(records_etfs_str[:1000])

<class 'pandas.core.frame.DataFrame'>
Index: 40 entries, SPY to SCHF
Data columns (total 31 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Dividend      38 non-null     float64
 1   Perf Quart    40 non-null     float64
 2   Perf Half     40 non-null     float64
 3   Beta          40 non-null     float64
 4   ATR           40 non-null     float64
 5   Volatility W  40 non-null     float64
 6   Volatility M  40 non-null     float64
 7   RSI           40 non-null     float64
 8   Avg Volume    40 non-null     float64
 9   Rel Volume    40 non-null     float64
 10  Expense       40 non-null     float64
 11  AUM           40 non-null     float64
 12  Flows% 1M     40 non-null     float64
 13  Flows% 3M     40 non-null     float64
 14  Flows% YTD    40 non-null     float64
 15  Return% 1Y    40 non-null     float64
 16  Return% 3Y    39 non-null     float64
 17  Return% 5Y    39 non-null     float64
 18  Tags          40 non-null     str

None

"[{'Ticker': 'SPY', 'Dividend': 7.06, 'Perf Quart': -0.0512, 'Perf Half': 0.0374, 'Beta': 1.01, 'ATR': 9.0, 'Volatility W': 0.0225, 'Volatility M': 0.0131, 'RSI': 34.03, 'Avg Volume': 51.73, 'Rel Volume': 1.53, 'Expense': 0.0009, 'AUM': 619.8, 'Flows% 1M': 0.0255, 'Flows% 3M': 0.0209, 'Flows% YTD': -0.0021, 'Return% 1Y': 0.1476, 'Return% 3Y': 0.1162, 'Return% 5Y': 0.1617, 'Tags': 'U.S., equity, SP500', 'Sharpe 30d': -3.3979008119880456, 'Sortino 30d': -3.990918671709931, 'Omega 30d': 0.5873202563164197, 'Sharpe 63d': -1.5902071609305564, 'Sortino 63d': -1.9749050315259344, 'Omega 63d': 0.7708935610829517, 'Sharpe 125d': 0.8524958725929831, 'Sortino 125d': 1.1702464905944368, 'Omega 125d': 1.1545307867220411, 'Sharpe 250d': 0.771301824252201, 'Sortino 250d': 1.056909545838279, 'Omega 250d': 1.1394362204487305}, {'Ticker': 'VOO', 'Dividend': 6.7, 'Perf Quart': -0.0511, 'Perf Half': 0.0377, 'Beta': 1.01, 'ATR': 8.29, 'Volatility W': 0.0225, 'Volatility M': 0.013, 'RSI': 34.17, 'Avg Volume

In [9]:
# Combine indexes from both dataframes into a single list of symbols
symbols = list(df_stocks.index) + list(df_etfs.index)
print(f"Total number of symbols: {len(symbols)}")
print(f"First 10 symbols: {symbols[:10]}")

Total number of symbols: 140
First 10 symbols: ['AAPL', 'MSFT', 'NVDA', 'AMZN', 'GOOGL', 'GOOG', 'META', 'BRK-A', 'BRK-B', 'TSM']


In [10]:
_df = pd.read_pickle('..\data\df_correlation_matrix.pkl')
display(_df)

df_correlation = _df.loc[symbols, symbols]
display(df_correlation)

# Convert DataFrame to list of dictionaries with index as 'Ticker'
records_correlation = df_correlation.reset_index().to_dict('records')

# Convert to string representation
records_correlation_str = str(records_correlation)

# Print first 1000 characters to preview
display(records_correlation_str[:1000])

Symbol      AAPL      NVDA      MSFT      AMZN      GOOG     GOOGL      META  \
Symbol                                                                         
AAPL    1.000000  0.865211  0.285015  0.625198  0.617776  0.617021  0.704202   
NVDA    0.865211  1.000000  0.397209  0.599437  0.617008  0.617164  0.632011   
MSFT    0.285015  0.397209  1.000000  0.286205  0.450420  0.456174  0.015666   
AMZN    0.625198  0.599437  0.286205  1.000000  0.796852  0.797941  0.832422   
GOOG    0.617776  0.617008  0.450420  0.796852  1.000000  0.999841  0.541711   
...          ...       ...       ...       ...       ...       ...       ...   
HYMB    0.868082  0.735108  0.035303  0.503809  0.321817  0.320690  0.737543   
CGMU    0.856385  0.673766  0.003453  0.468954  0.288031  0.286701  0.722494   
USRT    0.813389  0.711723  0.019319  0.401967  0.193074  0.192367  0.648267   
VFLO    0.696280  0.677029 -0.008725  0.792396  0.486420  0.484930  0.850610   
IWX     0.794393  0.741286 -0.022504  0.680532  0.418318  0.416195  0.849981   

Symbol      TSLA     BRK-B     BRK-A  ...      JAVA       EPI      BSCR  \
Symbol                                ...                                 
AAPL    0.774687  0.796218  0.786532  ...  0.807567  0.265630  0.893160   
NVDA    0.702289  0.652391  0.642806  ...  0.785788  0.258296  0.739450   
MSFT    0.261834 -0.129055 -0.134927  ...  0.058873  0.396537 -0.005601   
AMZN    0.906717  0.569965  0.569578  ...  0.726951 -0.351801  0.587875   
GOOG    0.742231  0.315257  0.310032  ...  0.482958 -0.081670  0.409732   
...          ...       ...       ...  ...       ...       ...       ...   
HYMB    0.644934  0.900310  0.892436  ...  0.873502  0.210867  0.958053   
CGMU    0.621624  0.890277  0.882221  ...  0.828467  0.215766  0.955374   
USRT    0.584943  0.898196  0.891321  ...  0.871524  0.304297  0.932673   
VFLO    0.839705  0.868264  0.868359  ...  0.956252 -0.224811  0.804598   
IWX     0.746654  0.953855  0.952216  ...  0.986500 -0.040198  0.914210   

Symbol      IBDS      JPIE      HYMB      CGMU      USRT      VFLO       IWX  
Symbol                                                                        
AAPL    0.893760  0.904572  0.868082  0.856385  0.813389  0.696280  0.794393  
NVDA    0.741677  0.758833  0.735108  0.673766  0.711723  0.677029  0.741286  
MSFT   -0.006358  0.010834  0.035303  0.003453  0.019319 -0.008725 -0.022504  
AMZN    0.587473  0.675452  0.503809  0.468954  0.401967  0.792396  0.680532  
GOOG    0.412213  0.503639  0.321817  0.288031  0.193074  0.486420  0.418318  
...          ...       ...       ...       ...       ...       ...       ...  
HYMB    0.957832  0.936670  1.000000  0.984041  0.944220  0.764925  0.893886  
CGMU    0.954831  0.929897  0.984041  1.000000  0.931876  0.729880  0.859495  
USRT    0.932250  0.891458  0.944220  0.931876  1.000000  0.743637  0.891519  
VFLO    0.804253  0.843486  0.764925  0.729880  0.743637  1.000000  0.938551  
IWX     0.914457  0.930398  0.893886  0.859495  0.891519  0.938551  1.000000  

[1394 rows x 1394 columns]

Symbol      AAPL      MSFT      NVDA      AMZN     GOOGL      GOOG      META  \
Symbol                                                                         
AAPL    1.000000  0.285015  0.865211  0.625198  0.617021  0.617776  0.704202   
MSFT    0.285015  1.000000  0.397209  0.286205  0.456174  0.450420  0.015666   
NVDA    0.865211  0.397209  1.000000  0.599437  0.617164  0.617008  0.632011   
AMZN    0.625198  0.286205  0.599437  1.000000  0.797941  0.796852  0.832422   
GOOGL   0.617021  0.456174  0.617164  0.797941  1.000000  0.999841  0.540764   
...          ...       ...       ...       ...       ...       ...       ...   
IBIT    0.520215  0.102403  0.521528  0.896297  0.642561  0.642586  0.761914   
QUAL    0.911905  0.267712  0.898181  0.721336  0.571285  0.571323  0.831628   
VCIT    0.831557  0.020434  0.666061  0.361547  0.208500  0.208881  0.639694   
TLT     0.265129  0.010976  0.126893 -0.373211 -0.400348 -0.401502 -0.056803   
SCHF    0.566061  0.025118  0.437158  0.193914  0.103969  0.106019  0.498146   

Symbol     BRK-A     BRK-B       TSM  ...       EFA      SPLG       IVW  \
Symbol                                ...                                 
AAPL    0.786532  0.796218  0.881991  ...  0.427074  0.902463  0.902734   
MSFT   -0.134927 -0.129055  0.330172  ... -0.004674  0.239042  0.314389   
NVDA    0.642806  0.652391  0.910990  ...  0.299114  0.876124  0.884514   
AMZN    0.569578  0.569965  0.792736  ...  0.089356  0.814720  0.859326   
GOOGL   0.307198  0.312437  0.731951  ...  0.039140  0.646322  0.735847   
...          ...       ...       ...  ...       ...       ...       ...   
IBIT    0.581046  0.579047  0.679080  ... -0.045865  0.744711  0.763165   
QUAL    0.860114  0.866007  0.926027  ...  0.487336  0.983999  0.959839   
VCIT    0.859737  0.867693  0.678168  ...  0.642197  0.775131  0.708558   
TLT     0.284027  0.291290 -0.005928  ...  0.540337  0.080940  0.001407   
SCHF    0.686338  0.685790  0.450745  ...  0.975509  0.549778  0.488755   

Symbol       XLF      SCHX      IBIT      QUAL      VCIT       TLT      SCHF  
Symbol                                                                        
AAPL    0.787066  0.891487  0.520215  0.911905  0.831557  0.265129  0.566061  
MSFT   -0.009845  0.232501  0.102403  0.267712  0.020434  0.010976  0.025118  
NVDA    0.720802  0.867291  0.521528  0.898181  0.666061  0.126893  0.437158  
AMZN    0.818555  0.827917  0.896297  0.721336  0.361547 -0.373211  0.193914  
GOOGL   0.544963  0.651694  0.642561  0.571285  0.208500 -0.400348  0.103969  
...          ...       ...       ...       ...       ...       ...       ...  
IBIT    0.819073  0.762984  1.000000  0.640064  0.294161 -0.417430  0.058753  
QUAL    0.903164  0.978884  0.640064  1.000000  0.835609  0.223474  0.631763  
VCIT    0.719254  0.760767  0.294161  0.835609  1.000000  0.646298  0.767860  
TLT    -0.012280  0.057718 -0.417430  0.223474  0.646298  1.000000  0.566031  
SCHF    0.488936  0.533400  0.058753  0.631763  0.767860  0.566031  1.000000  

[140 rows x 140 columns]

"[{'Symbol': 'AAPL', 'AAPL': 1.0, 'MSFT': 0.28501454451248814, 'NVDA': 0.8652105003269625, 'AMZN': 0.6251976221437099, 'GOOGL': 0.6170209701155636, 'GOOG': 0.617776162831828, 'META': 0.7042015985024415, 'BRK-A': 0.786531524561205, 'BRK-B': 0.7962180091811142, 'TSM': 0.8819912967665464, 'LLY': 0.4508499204265386, 'TSLA': 0.7746869814106482, 'AVGO': 0.7992033214297684, 'WMT': 0.8547998274201383, 'JPM': 0.7566765201218123, 'V': 0.5385719693216026, 'MA': 0.6231186392442746, 'XOM': -0.14107299357224415, 'COST': 0.8926990399351875, 'UNH': 0.4990941354597536, 'ORCL': 0.8149525824028736, 'PG': 0.7135122768486154, 'JNJ': 0.3067717146111126, 'NFLX': 0.7530128324547767, 'HD': 0.6738068163392149, 'ABBV': 0.5625764235531812, 'BABA': 0.5737759918329772, 'SAP': 0.795312224989895, 'BAC': 0.7979617944028115, 'KO': 0.5805445837093942, 'TMUS': 0.8164559846532998, 'NVO': -0.5385207255739397, 'ASML': -0.6239493684874367, 'CRM': 0.3659565439543356, 'CVX': -0.1649909203728438, 'CSCO': 0.6786041206459721, 'TM

In [11]:
# Create the formatted string with triple quotes
formatted_records = f'records_stocks = """{records_stocks_str}"""'

# Save to a file in the data directory
output_path = Path(ROOT_DIR) / 'data' / 'records_stocks.txt'
with open(output_path, 'w') as f:
  f.write(formatted_records)

print(f"File saved to: {output_path}")

File saved to: c:\Users\ping\Files_win10\python\py310\stocks\data\records_stocks.txt


In [12]:
# Create the formatted string with triple quotes
formatted_records = f'records_etfs = """{records_etfs_str}"""'

# Save to a file in the data directory
output_path = Path(ROOT_DIR) / 'data' / 'records_etfs.txt'
with open(output_path, 'w') as f:
  f.write(formatted_records)

print(f"File saved to: {output_path}")

File saved to: c:\Users\ping\Files_win10\python\py310\stocks\data\records_etfs.txt


In [13]:
# Create the formatted string with triple quotes
formatted_records = f'records_correlation = """{records_correlation_str}"""'

# Save to a file in the data directory
output_path = Path(ROOT_DIR) / 'data' / 'records_correlation.txt'
with open(output_path, 'w') as f:
  f.write(formatted_records)

print(f"File saved to: {output_path}")

File saved to: c:\Users\ping\Files_win10\python\py310\stocks\data\records_correlation.txt


you are a financial expert in stock trading, stock analysis, and building and managing financial portfolios.
you are a financial expert in selecting stocks and ETFs to build a portfolio. your goal is to build a  portfolio with stocks and ETFs combination that will most likely to produce superior returns with acceptable risk for the each of the next 15, 30, 63, 125 trading days. user will give you financial data for stocks, ETFs and their Sharpe Ratio, Sortino Ratio, and Omega Ratio for the past 30, 63, 125, 250 trading days. User will also give you the correlation matrix of these stocks and ETFs for the past 250 trading days. your job is to select the stocks and ETFs for portfolio to produce superior returns with acceptable risk for the each of the next 15, 30, 63, 125 trading days. Please provide the allocation percentage for each symbols in the portfolio.

Here is the user profile:
Investor-Specific Risk Aversion Coefficient: 1.3 where 0 is aggressive and 5 is risk-averse
Transaction Costs: 0.3%
Tax: 0%
Liquidity Needs: no need for future cash



i will give you stocks, ETFs, correlation data. don't analyze the data, just look at it, and tell me do you understand the data. do you need clarification? 

In [18]:
# Load your DataFrame
df = pd.read_pickle('../data/df_finviz_stocks_n_ratios.pkl')  # Use raw string or double backslashes

In [ ]:
# AI Studio

import pandas as pd
import numpy as np
from scipy.stats import zscore
from pypfopt import EfficientFrontier, objective_functions
from pypfopt import risk_models
from pypfopt import expected_returns

def select_outperforming_stocks(df):
    """Stock selection with debug output"""
    print("\n=== Starting Stock Selection ===")
    
    # Handle Ticker position
    if df.index.name == 'Ticker':
        temp_df = df.reset_index()
        print("Reset Ticker index to column")
    else:
        temp_df = df.copy()
        print("Ticker already in columns")

    # Validate required columns
    required_cols = {
        'Ticker', 'Sharpe 30d', 'Sortino 30d', 'Omega 30d',
        'Sharpe 63d', 'Sortino 63d', 'Omega 63d', 'Sharpe 125d',
        'Sortino 125d', 'Omega 125d', 'EPS next Y', 'Profit M',
        'PEG', 'Beta', 'Inst Own', 'Recom', 'Volatility M'
    }
    missing = required_cols - set(temp_df.columns)
    if missing:
        raise ValueError(f"Missing columns: {missing}")
    
    print(f"Processing {len(temp_df)} stocks")

    # Horizon configuration
    horizons = {
        15: '30d',
        30: '30d', 
        60: '63d',
        120: '125d'
    }

    selected_stocks = {}
    
    for days, suffix in horizons.items():
        print(f"\n-- Analyzing {days}-day horizon --")
        sharpe = f'Sharpe {suffix}'
        sortino = f'Sortino {suffix}'
        omega = f'Omega {suffix}'
        
        # Filter stocks
        valid = temp_df[
            (temp_df[sharpe] > 0) &
            (temp_df[sortino] > 0) &
            (temp_df[omega] > 1)
        ].copy()
        
        print(f"Found {len(valid)} qualifying stocks")
        
        if valid.empty:
            selected_stocks[f'{days}_days'] = []
            continue

        # Scoring metrics
        scoring = [
            (sharpe, 1.5, False),
            (sortino, 1.5, False),
            (omega, 1.2, False),
            ('EPS next Y', 1.0, False),
            ('Profit M', 0.8, False),
            ('PEG', 0.7, True),
            ('Beta', 0.7, True),
            ('Inst Own', 0.5, False),
            ('Recom', 0.5, True)
        ]

        # Calculate scores
        for metric, weight, invert in scoring:
            z_vals = zscore(valid[metric], nan_policy='omit')
            z_vals = np.nan_to_num(z_vals, nan=0.0)
            if invert:
                z_vals = -z_vals
            valid[f'{metric}_score'] = z_vals * weight

        # Total score
        score_cols = [f'{metric}_score' for metric, _, _ in scoring]
        valid['total_score'] = valid[score_cols].sum(axis=1)
        
        # # Select top 5
        # top_5 = valid.nlargest(5, 'total_score')['Ticker'].tolist()
        # print(f"Selected top 5: {top_5}")
        # selected_stocks[f'{days}_days'] = top_5[:5]

        # Select top 100
        top_100 = valid.nlargest(100, 'total_score')['Ticker'].tolist()
        print(f"Selected top 100: {top_100}")
        selected_stocks[f'{days}_days'] = top_100[:100]


    return selected_stocks

def optimize_portfolio(selected_tickers, df_metrics, df_correlation):
    """Portfolio optimization with debug output"""
    print("\n=== Starting Portfolio Optimization ===")
    
    # Handle Ticker index
    if df_metrics.index.name == 'Ticker':
        metrics = df_metrics
        print("Ticker index already set in metrics")
    else:
        metrics = df_metrics.set_index('Ticker')
        print("Set Ticker index in metrics")

    # Common tickers handling
    common = list(set(selected_tickers) 
                  & set(df_correlation.index) 
                  & set(metrics.index))
    
    print(f"Common tickers: {common}")
    
    if not common:
        raise ValueError("No common tickers between selection and correlation matrix")
    
    # Prepare data
    df_corr = df_correlation.loc[common, common]
    df_sub = metrics.loc[common]
    
    # Calculate covariance matrix
    print("\nCalculating covariance matrix...")
    #volatilities = df_sub['Volatility M'] * np.sqrt(21)  # Monthly to annualized
    #cov_matrix = pd.DataFrame(
    #    np.outer(volatilities, volatilities) * df_corr.values,
    #    index=common,
    #    columns=common
    #)

    # Use sample covariance matrix directly from returns (more robust)
    returns_data = df_metrics.loc[common, ['Sharpe 30d', 'Sharpe 63d', 'Sharpe 125d']].fillna(0) # Replace nan with 0, but a better approach could use more sophisticated imputation
    sample_cov = risk_models.sample_cov(returns_data)
    cov_matrix = pd.DataFrame(sample_cov, index=common, columns=common) # ensure correct indexing

    
    # Calculate expected returns
    print("Calculating expected returns...")
    #df_sub['Expected Return'] = (
    #    0.4 * df_sub['EPS next Y'] +
    #    0.3 * df_sub['Profit M'] +
    #    0.2 * df_sub['Sales Q/Q'] +
    #    0.1 * (5 - df_sub['Recom'])  # Inverse since lower Recom is better
    #)
    
    # Using mean historical return (less prone to errors)
    #mu = expected_returns.mean_historical_return(df_metrics.loc[common, ['Sharpe 30d', 'Sharpe 63d', 'Sharpe 125d']].fillna(0))  # Replace nan with 0 for calculation
    # expected_returns = mu
    #df_sub['Expected Return'] = expected_returns
    
    #Using the mean of Sharpe Ratios as a proxy of expected returns
    df_sub['Expected Return'] = df_sub[['Sharpe 30d', 'Sharpe 63d', 'Sharpe 125d']].mean(axis=1)
    expected_returns = df_sub['Expected Return'].values  # Convert to numpy array
    expected_returns = pd.Series(expected_returns, index=common)  # Ensure correct indexing


    # Portfolio optimization
    print("Optimizing portfolio...")
    ef = EfficientFrontier(
        expected_returns=expected_returns, # Use the correctly indexed series
        cov_matrix=cov_matrix,
        weight_bounds=(0.02, 0.25)
    )
    ef.add_objective(objective_functions.L2_reg, gamma=0.1)
    try:
        ef.max_sharpe()
    except Exception as e:
        print(f"Optimization failed: {e}")
        return {}  # Or some other appropriate handling

    
    weights = ef.clean_weights()
    print("\n=== Optimization Results ===")
    print("Portfolio weights:")
    for ticker, weight in weights.items():
        print(f"{ticker}: {weight:.2%}")
    
    return weights

def full_analysis_pipeline(df_metrics, df_correlation):
    """Complete pipeline with debug output"""
    print("=== Starting Full Analysis Pipeline ===")
    
    # Ensure metrics has Ticker index
    if df_metrics.index.name != 'Ticker':
        df_metrics = df_metrics.set_index('Ticker')
        print("Set Ticker index in metrics DataFrame")
    
    # Step 1: Stock selection
    selected = select_outperforming_stocks(df_metrics)
    all_tickers = list(set(
        t for horizon in selected.values() 
        for t in horizon if isinstance(t, str)
        ))
    
    print(f"\nFinal selected tickers: {all_tickers}")
    
    # Step 2: Portfolio optimization
    if not all_tickers:
        raise ValueError("No stocks selected in initial filtering")
        
    return optimize_portfolio(all_tickers, df_metrics, df_correlation)

# Usage example:
# df_metrics = pd.read_pickle(r'..\data\df_finviz_stocks_n_ratios.pkl')
# df_correlation = pd.read_csv('correlation_matrix.csv', index_col=0)
# portfolio = full_analysis_pipeline(df_metrics, df_correlation)

df_metrics = pd.read_pickle('..\data\df_finviz_stocks_n_ratios.pkl')

_df = pd.read_pickle('..\data\df_correlation_matrix.pkl')
df_correlation = _df.loc[symbols, symbols]

portfolio = full_analysis_pipeline(df_metrics, df_correlation)
portfolio

=== Starting Full Analysis Pipeline ===

=== Starting Stock Selection ===
Reset Ticker index to column
Processing 1000 stocks

-- Analyzing 15-day horizon --
Found 423 qualifying stocks
Selected top 5: ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN']

-- Analyzing 30-day horizon --
Found 423 qualifying stocks
Selected top 5: ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN']

-- Analyzing 60-day horizon --
Found 379 qualifying stocks
Selected top 5: ['ITCI', 'ESLT', 'VRSN', 'BBVA', 'ONC']

-- Analyzing 120-day horizon --
Found 623 qualifying stocks
Selected top 5: ['ESLT', 'AFRM', 'ITCI', 'GILD', 'UNM']

Final selected tickers: ['AFRM', 'ESLT', 'BBVA', 'VRSN', 'ITCI', 'BUD', 'ONC', 'UNM', 'GILD', 'ABBV', 'YUM']

=== Starting Portfolio Optimization ===
Ticker index already set in metrics
Common tickers: ['GILD', 'ABBV']

Calculating covariance matrix...
Calculating expected returns...
Optimizing portfolio...
ERROR in LDL_factor: Error in KKT matrix LDL factorization when computing the nonzero elements. The p

c:\Users\ping\Files_win10\python\py310\.venv\lib\site-packages\pandas\core\frame.py:11211: RuntimeWarning: Degrees of freedom <= 0 for slice
  base_cov = np.cov(mat.T, ddof=ddof)
c:\Users\ping\Files_win10\python\py310\.venv\lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning: divide by zero encountered in divide
  c *= np.true_divide(1, fact)
c:\Users\ping\Files_win10\python\py310\.venv\lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning: invalid value encountered in multiply
  c *= np.true_divide(1, fact)
c:\Users\ping\Files_win10\python\py310\.venv\lib\site-packages\pypfopt\efficient_frontier\efficient_frontier.py:259: UserWarning: max_sharpe transforms the optimization problem so additional objectives may not work as expected.
  warnings.warn(


{}

In [53]:
# AI Studio

import pandas as pd
import numpy as np
from scipy.stats import zscore
from pypfopt import EfficientFrontier, objective_functions
from pypfopt import risk_models
from pypfopt import expected_returns

def select_outperforming_stocks(df):
    """Stock selection with debug output"""
    print("\n=== Starting Stock Selection ===")
    
    # Handle Ticker position
    if df.index.name == 'Ticker':
        temp_df = df.reset_index()
        print("Reset Ticker index to column")
    else:
        temp_df = df.copy()
        print("Ticker already in columns")

    # Validate required columns
    required_cols = {
        'Ticker', 'Sharpe 30d', 'Sortino 30d', 'Omega 30d',
        'Sharpe 63d', 'Sortino 63d', 'Omega 63d', 'Sharpe 125d',
        'Sortino 125d', 'Omega 125d', 'EPS next Y', 'Profit M',
        'PEG', 'Beta', 'Inst Own', 'Recom', 'Volatility M'
    }
    missing = required_cols - set(temp_df.columns)
    if missing:
        raise ValueError(f"Missing columns: {missing}")
    
    print(f"Processing {len(temp_df)} stocks")

    # Horizon configuration
    horizons = {
        15: '30d',
        30: '30d', 
        60: '63d',
        120: '125d'
    }

    selected_stocks = {}
    
    for days, suffix in horizons.items():
        print(f"\n-- Analyzing {days}-day horizon --")
        sharpe = f'Sharpe {suffix}'
        sortino = f'Sortino {suffix}'
        omega = f'Omega {suffix}'
        
        # Filter stocks
        valid = temp_df[
            (temp_df[sharpe] > 0) &
            (temp_df[sortino] > 0) &
            (temp_df[omega] > 1)
        ].copy()
        
        print(f"Found {len(valid)} qualifying stocks")
        
        if valid.empty:
            selected_stocks[f'{days}_days'] = []
            continue

        # Scoring metrics
        scoring = [
            (sharpe, 1.5, False),
            (sortino, 1.5, False),
            (omega, 1.2, False),
            ('EPS next Y', 1.0, False),
            ('Profit M', 0.8, False),
            ('PEG', 0.7, True),
            ('Beta', 0.7, True),
            ('Inst Own', 0.5, False),
            ('Recom', 0.5, True)
        ]

        # Calculate scores
        for metric, weight, invert in scoring:
            z_vals = zscore(valid[metric], nan_policy='omit')
            z_vals = np.nan_to_num(z_vals, nan=0.0)
            if invert:
                z_vals = -z_vals
            valid[f'{metric}_score'] = z_vals * weight

        # Total score
        score_cols = [f'{metric}_score' for metric, _, _ in scoring]
        valid['total_score'] = valid[score_cols].sum(axis=1)
        
        # Select top 5
        top_5 = valid.nlargest(5, 'total_score')['Ticker'].tolist()
        print(f"Selected top 5: {top_5}")
        selected_stocks[f'{days}_days'] = top_5[:5]

    return selected_stocks

def optimize_portfolio(selected_tickers, df_metrics, df_correlation):
    """Portfolio optimization with debug output"""
    print("\n=== Starting Portfolio Optimization ===")
    
    # Handle Ticker index
    if df_metrics.index.name == 'Ticker':
        metrics = df_metrics
        print("Ticker index already set in metrics")
    else:
        metrics = df_metrics.set_index('Ticker')
        print("Set Ticker index in metrics")

    # Common tickers handling
    common = list(set(selected_tickers) 
                  & set(df_correlation.index) 
                  & set(metrics.index))
    
    print(f"Common tickers: {common}")
    
    if not common:
        raise ValueError("No common tickers between selection and correlation matrix")
    
    # Prepare data
    df_corr = df_correlation.loc[common, common]
    df_sub = metrics.loc[common]
    
    # Calculate covariance matrix
    print("\nCalculating covariance matrix...")
    #volatilities = df_sub['Volatility M'] * np.sqrt(21)  # Monthly to annualized
    #cov_matrix = pd.DataFrame(
    #    np.outer(volatilities, volatilities) * df_corr.values,
    #    index=common,
    #    columns=common
    #)

    # Use sample covariance matrix directly from returns (more robust)
    returns_data = df_metrics.loc[common, ['Sharpe 30d', 'Sharpe 63d', 'Sharpe 125d']].fillna(0) # Replace nan with 0, but a better approach could use more sophisticated imputation
    sample_cov = risk_models.sample_cov(returns_data)
    cov_matrix = sample_cov


    
    # Calculate expected returns
    print("Calculating expected returns...")
    #df_sub['Expected Return'] = (
    #    0.4 * df_sub['EPS next Y'] +
    #    0.3 * df_sub['Profit M'] +
    #    0.2 * df_sub['Sales Q/Q'] +
    #    0.1 * (5 - df_sub['Recom'])  # Inverse since lower Recom is better
    #)
    
    # Using mean historical return (less prone to errors)
    #mu = expected_returns.mean_historical_return(df_metrics.loc[common, ['Sharpe 30d', 'Sharpe 63d', 'Sharpe 125d']].fillna(0))  # Replace nan with 0 for calculation
    # expected_returns = mu
    #df_sub['Expected Return'] = expected_returns
    
    #Using the mean of Sharpe Ratios as a proxy of expected returns
    df_sub['Expected Return'] = df_sub[['Sharpe 30d', 'Sharpe 63d', 'Sharpe 125d']].mean(axis=1)


    # Portfolio optimization
    print("Optimizing portfolio...")
    ef = EfficientFrontier(
        expected_returns=df_sub['Expected Return'],
        cov_matrix=cov_matrix,
        weight_bounds=(0.02, 0.25)
    )
    ef.add_objective(objective_functions.L2_reg, gamma=0.1)
    try:
        ef.max_sharpe()
    except Exception as e:
        print(f"Optimization failed: {e}")
        return {}  # Or some other appropriate handling

    
    weights = ef.clean_weights()
    print("\n=== Optimization Results ===")
    print("Portfolio weights:")
    for ticker, weight in weights.items():
        print(f"{ticker}: {weight:.2%}")
    
    return weights

def full_analysis_pipeline(df_metrics, df_correlation):
    """Complete pipeline with debug output"""
    print("=== Starting Full Analysis Pipeline ===")
    
    # Ensure metrics has Ticker index
    if df_metrics.index.name != 'Ticker':
        df_metrics = df_metrics.set_index('Ticker')
        print("Set Ticker index in metrics DataFrame")
    
    # Step 1: Stock selection
    selected = select_outperforming_stocks(df_metrics)
    all_tickers = list(set(
        t for horizon in selected.values() 
        for t in horizon if isinstance(t, str)
        ))
    
    print(f"\nFinal selected tickers: {all_tickers}")
    
    # Step 2: Portfolio optimization
    if not all_tickers:
        raise ValueError("No stocks selected in initial filtering")
        
    return optimize_portfolio(all_tickers, df_metrics, df_correlation)

# Usage example:
# df_metrics = pd.read_pickle(r'..\data\df_finviz_stocks_n_ratios.pkl')
# df_correlation = pd.read_csv('correlation_matrix.csv', index_col=0)
# portfolio = full_analysis_pipeline(df_metrics, df_correlation)

df_metrics = pd.read_pickle('..\data\df_finviz_stocks_n_ratios.pkl')

_df = pd.read_pickle('..\data\df_correlation_matrix.pkl')
df_correlation = _df.loc[symbols, symbols]

portfolio = full_analysis_pipeline(df_metrics, df_correlation)
portfolio

=== Starting Full Analysis Pipeline ===

=== Starting Stock Selection ===
Reset Ticker index to column
Processing 1000 stocks

-- Analyzing 15-day horizon --
Found 423 qualifying stocks
Selected top 5: ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN']

-- Analyzing 30-day horizon --
Found 423 qualifying stocks
Selected top 5: ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN']

-- Analyzing 60-day horizon --
Found 379 qualifying stocks
Selected top 5: ['ITCI', 'ESLT', 'VRSN', 'BBVA', 'ONC']

-- Analyzing 120-day horizon --
Found 623 qualifying stocks
Selected top 5: ['ESLT', 'AFRM', 'ITCI', 'GILD', 'UNM']

Final selected tickers: ['AFRM', 'ESLT', 'BBVA', 'VRSN', 'ITCI', 'BUD', 'ONC', 'UNM', 'GILD', 'ABBV', 'YUM']

=== Starting Portfolio Optimization ===
Ticker index already set in metrics
Common tickers: ['GILD', 'ABBV']

Calculating covariance matrix...


c:\Users\ping\Files_win10\python\py310\.venv\lib\site-packages\pandas\core\frame.py:11211: RuntimeWarning: Degrees of freedom <= 0 for slice
  base_cov = np.cov(mat.T, ddof=ddof)
c:\Users\ping\Files_win10\python\py310\.venv\lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning: divide by zero encountered in divide
  c *= np.true_divide(1, fact)
c:\Users\ping\Files_win10\python\py310\.venv\lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning: invalid value encountered in multiply
  c *= np.true_divide(1, fact)


Calculating expected returns...
Optimizing portfolio...


ValueError: Covariance matrix does not match expected returns

In [69]:
# DeepSeek

import pandas as pd
import numpy as np
from scipy.stats import zscore
from pypfopt import EfficientFrontier, objective_functions

def select_outperforming_stocks(df):
    """Stock selection with debug output"""
    print("\n=== Starting Stock Selection ===")
    
    # Handle Ticker position
    if df.index.name == 'Ticker':
        temp_df = df.reset_index()
        print("Reset Ticker index to column")
    else:
        temp_df = df.copy()
        print("Ticker already in columns")

    # Validate required columns
    required_cols = {
        'Ticker', 'Sharpe 30d', 'Sortino 30d', 'Omega 30d',
        'Sharpe 63d', 'Sortino 63d', 'Omega 63d', 'Sharpe 125d',
        'Sortino 125d', 'Omega 125d', 'EPS next Y', 'Profit M',
        'PEG', 'Beta', 'Inst Own', 'Recom', 'Volatility M'
    }
    missing = required_cols - set(temp_df.columns)
    if missing:
        raise ValueError(f"Missing columns: {missing}")
    
    print(f"Processing {len(temp_df)} stocks")

    # Horizon configuration
    horizons = {
        15: '30d',
        30: '30d', 
        60: '63d',
        120: '125d'
    }

    selected_stocks = {}
    
    for days, suffix in horizons.items():
        print(f"\n-- Analyzing {days}-day horizon --")
        sharpe = f'Sharpe {suffix}'
        sortino = f'Sortino {suffix}'
        omega = f'Omega {suffix}'
        
        # Filter stocks
        valid = temp_df[
            (temp_df[sharpe] > 0) &
            (temp_df[sortino] > 0) &
            (temp_df[omega] > 1)
        ].copy()
        
        print(f"Found {len(valid)} qualifying stocks")
        
        if valid.empty:
            selected_stocks[f'{days}_days'] = []
            continue

        # Scoring metrics
        scoring = [
            (sharpe, 1.5, False),
            (sortino, 1.5, False),
            (omega, 1.2, False),
            ('EPS next Y', 1.0, False),
            ('Profit M', 0.8, False),
            ('PEG', 0.7, True),
            ('Beta', 0.7, True),
            ('Inst Own', 0.5, False),
            ('Recom', 0.5, True)
        ]

        # Calculate scores
        for metric, weight, invert in scoring:
            z_vals = zscore(valid[metric], nan_policy='omit')
            z_vals = np.nan_to_num(z_vals, nan=0.0)
            if invert:
                z_vals = -z_vals
            valid[f'{metric}_score'] = z_vals * weight

        # Total score
        score_cols = [f'{metric}_score' for metric, _, _ in scoring]
        valid['total_score'] = valid[score_cols].sum(axis=1)
        
        # # Select top 5
        # top_5 = valid.nlargest(5, 'total_score')['Ticker'].tolist()
        # print(f"Selected top 5: {top_5}")
        # selected_stocks[f'{days}_days'] = top_5[:5]

        # Select top 100
        top_100 = valid.nlargest(100, 'total_score')['Ticker'].tolist()
        print(f"Selected top 100: {top_100}")
        selected_stocks[f'{days}_days'] = top_100[:100]


    return selected_stocks

def optimize_portfolio(selected_tickers, df_metrics, df_correlation):
    """Portfolio optimization with debug output"""
    print("\n=== Starting Portfolio Optimization ===")
    
    # Handle Ticker index
    if df_metrics.index.name == 'Ticker':
        metrics = df_metrics
        print("Ticker index already set in metrics")
    else:
        metrics = df_metrics.set_index('Ticker')
        print("Set Ticker index in metrics")

    # Common tickers handling
    common = list(set(selected_tickers) 
                  & set(df_correlation.index) 
                  & set(metrics.index))
    
    print(f"Common tickers: {common}")
    
    if not common:
        raise ValueError("No common tickers between selection and correlation matrix")
    
    # Prepare data
    df_corr = df_correlation.loc[common, common]
    df_sub = metrics.loc[common]
    
    # Calculate covariance matrix
    print("\nCalculating covariance matrix...")
    volatilities = df_sub['Volatility M'] * np.sqrt(21)  # Monthly to annualized
    cov_matrix = pd.DataFrame(
        np.outer(volatilities, volatilities) * df_corr.values,
        index=common,
        columns=common
    )
    
    # Calculate expected returns
    print("Calculating expected returns...")
    df_sub['Expected Return'] = (
        0.4 * df_sub['EPS next Y'] +
        0.3 * df_sub['Profit M'] +
        0.2 * df_sub['Sales Q/Q'] +
        0.1 * (5 - df_sub['Recom'])  # Inverse since lower Recom is better
    )
    
    # Portfolio optimization
    print("Optimizing portfolio...")
    ef = EfficientFrontier(
        expected_returns=df_sub['Expected Return'],
        cov_matrix=cov_matrix,
        weight_bounds=(0.02, 0.25)
    )
    ef.add_objective(objective_functions.L2_reg, gamma=0.1)
    ef.max_sharpe()
    
    weights = ef.clean_weights()
    print("\n=== Optimization Results ===")
    print("Portfolio weights:")
    for ticker, weight in weights.items():
        print(f"{ticker}: {weight:.2%}")
    
    return weights

def full_analysis_pipeline(df_metrics, df_correlation):
    """Complete pipeline with debug output"""
    print("=== Starting Full Analysis Pipeline ===")
    
    # Ensure metrics has Ticker index
    if df_metrics.index.name != 'Ticker':
        df_metrics = df_metrics.set_index('Ticker')
        print("Set Ticker index in metrics DataFrame")
    
    # Step 1: Stock selection
    selected = select_outperforming_stocks(df_metrics)
    all_tickers = list(set(
        t for horizon in selected.values() 
        for t in horizon if isinstance(t, str)
        ))
    
    print(f"\nFinal selected tickers: {all_tickers}")
    
    # Step 2: Portfolio optimization
    if not all_tickers:
        raise ValueError("No stocks selected in initial filtering")
        
    return optimize_portfolio(all_tickers, df_metrics, df_correlation)

# Usage example:
# df_metrics = pd.read_pickle(r'..\data\df_finviz_stocks_n_ratios.pkl')
# df_correlation = pd.read_csv('correlation_matrix.csv', index_col=0)
# portfolio = full_analysis_pipeline(df_metrics, df_correlation)

df_metrics = pd.read_pickle('..\data\df_finviz_stocks_n_ratios.pkl')

_df = pd.read_pickle('..\data\df_correlation_matrix.pkl')
df_correlation = _df.loc[symbols, symbols]

portfolio = full_analysis_pipeline(df_metrics, df_correlation)
portfolio


=== Starting Full Analysis Pipeline ===

=== Starting Stock Selection ===
Reset Ticker index to column
Processing 1000 stocks

-- Analyzing 15-day horizon --
Found 423 qualifying stocks
Selected top 100: ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN', 'GILD', 'PGR', 'ABEV', 'TEF', 'WTRG', 'ONC', 'SAN', 'BEKE', 'EXEL', 'TAK', 'ING', 'BABA', 'LYG', 'AZN', 'JNJ', 'VZ', 'BMRN', 'WPC', 'T', 'BBVA', 'ESLT', 'TMUS', 'AWK', 'EA', 'CMS', 'MDLZ', 'AMGN', 'PM', 'MCD', 'HSY', 'KO', 'KMB', 'BRO', 'CHKP', 'XPEV', 'WPM', 'TU', 'CVS', 'VICI', 'UMC', 'SNN', 'ED', 'XP', 'NVS', 'EXC', 'MT', 'ECL', 'KR', 'PUK', 'EGP', 'TTWO', 'NLY', 'SBAC', 'INVH', 'AON', 'AMT', 'AGNC', 'FMX', 'HALO', 'DTE', 'JAZZ', 'SNY', 'BJ', 'RSG', 'ASND', 'DGX', 'LKQ', 'EBR', 'GSK', 'FOXA', 'MAA', 'FR', 'HSBC', 'AZO', 'MNST', 'UDR', 'SONY', 'QSR', 'RBA', 'AJG', 'WMG', 'WM', 'IBM', 'MCK', 'RPRX', 'WEC', 'BCH', 'VRTX', 'PNW', 'VTR', 'GOLD', 'SJM', 'STE', 'CPT', 'VOD']

-- Analyzing 30-day horizon --
Found 423 qualifying stocks
Selected top 100:

c:\Users\ping\Files_win10\python\py310\.venv\lib\site-packages\pypfopt\efficient_frontier\efficient_frontier.py:259: UserWarning: max_sharpe transforms the optimization problem so additional objectives may not work as expected.
  warnings.warn(


SolverError: Workspace allocation error!

In [68]:
if df_metrics.index.name == 'Ticker':
    metrics = df_metrics

common = list(set(all_tickers) 
              & set(df_correlation.index) 
              & set(metrics.index))

print(f'set(all_tickers) {len(all_tickers)}: {set(all_tickers)}')
print(f'set(df_correlation.index) {len(df_correlation.index)}: {set(df_correlation.index)}')
print(f'set(metrics.index) {len(metrics.index)}: {set(metrics.index)}')
print(f"Common tickers {len(common)}: {common}")

set(all_tickers) 24: {'AFRM', 'GILD', 'ABBV', 'YUM', 'ABEV', 'SAN', 'ESLT', 'LPLA', 'VRSN', 'LYG', 'BBVA', 'YMM', 'BUD', 'BABA', 'PGR', 'UNM', 'TEF', 'BJ', 'WTRG', 'CSCO', 'ITCI', 'ING', 'ONC', 'TPR'}
set(df_correlation.index) 140: {'CRM', 'C', 'GOOG', 'VUG', 'AVGO', 'TJX', 'SPLG', 'LLY', 'WFC', 'BLK', 'NVDA', 'MS', 'AGG', 'HSBC', 'IVV', 'NFLX', 'VYM', 'JNJ', 'PGR', 'NVO', 'XLF', 'SCHF', 'NOW', 'IEMG', 'VXUS', 'VOO', 'MRK', 'VZ', 'IEFA', 'ORCL', 'SHEL', 'BKNG', 'NEE', 'CMCSA', 'PFE', 'TMUS', 'PDD', 'UBER', 'AXP', 'CVX', 'GS', 'ISRG', 'BRK-B', 'NVS', 'BSX', 'WMT', 'DIS', 'ACN', 'UL', 'SONY', 'BABA', 'DHR', 'HD', 'BAC', 'AZN', 'AMGN', 'TTE', 'SAP', 'JPM', 'XLK', 'VO', 'VEA', 'BRK-A', 'IWD', 'META', 'DE', 'AAPL', 'VWO', 'UNP', 'V', 'UNH', 'TM', 'ITOT', 'AMD', 'CAT', 'GILD', 'VGT', 'BX', 'SCHX', 'GE', 'SYK', 'BHP', 'IVW', 'AMZN', 'SNY', 'IJR', 'TLT', 'QCOM', 'SHOP', 'PLTR', 'IWF', 'IBIT', 'VRTX', 'ASML', 'GOOGL', 'RTX', 'INTU', 'ADBE', 'BND', 'IJH', 'LOW', 'EFA', 'MSFT', 'VB', 'SPGI', 'IBM

In [61]:
selected = select_outperforming_stocks(df_metrics)
all_tickers = list(set(
    t for horizon in selected.values() 
    for t in horizon if isinstance(t, str)
    ))

print(f"\nFinal selected tickers {len(all_tickers)}: {all_tickers}")


=== Starting Stock Selection ===
Reset Ticker index to column
Processing 1000 stocks

-- Analyzing 15-day horizon --
Found 423 qualifying stocks
Selected top 10: ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN', 'GILD', 'PGR', 'ABEV', 'TEF', 'WTRG']

-- Analyzing 30-day horizon --
Found 423 qualifying stocks
Selected top 10: ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN', 'GILD', 'PGR', 'ABEV', 'TEF', 'WTRG']

-- Analyzing 60-day horizon --
Found 379 qualifying stocks
Selected top 10: ['ITCI', 'ESLT', 'VRSN', 'BBVA', 'ONC', 'GILD', 'SAN', 'LYG', 'ING', 'BABA']

-- Analyzing 120-day horizon --
Found 623 qualifying stocks
Selected top 10: ['ESLT', 'AFRM', 'ITCI', 'GILD', 'UNM', 'TPR', 'LPLA', 'CSCO', 'YMM', 'BJ']

Final selected tickers 24: ['AFRM', 'GILD', 'ABBV', 'YUM', 'ABEV', 'SAN', 'ESLT', 'LPLA', 'VRSN', 'LYG', 'BBVA', 'YMM', 'BUD', 'BABA', 'PGR', 'UNM', 'TEF', 'BJ', 'WTRG', 'CSCO', 'ITCI', 'ING', 'ONC', 'TPR']


In [50]:
df_metrics = pd.read_pickle('..\data\df_finviz_stocks_n_ratios.pkl')

In [51]:
_df = pd.read_pickle('..\data\df_correlation_matrix.pkl')
df_correlation = _df.loc[symbols, symbols]

In [52]:
portfolio = full_analysis_pipeline(df_metrics, df_correlation)
portfolio

=== Starting Full Analysis Pipeline ===

=== Starting Stock Selection ===
Reset Ticker index to column
Processing 1000 stocks

-- Analyzing 15-day horizon --
Found 423 qualifying stocks
Selected top 5: ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN']

-- Analyzing 30-day horizon --
Found 423 qualifying stocks
Selected top 5: ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN']

-- Analyzing 60-day horizon --
Found 379 qualifying stocks
Selected top 5: ['ITCI', 'ESLT', 'VRSN', 'BBVA', 'ONC']

-- Analyzing 120-day horizon --
Found 623 qualifying stocks
Selected top 5: ['ESLT', 'AFRM', 'ITCI', 'GILD', 'UNM']

Final selected tickers: ['AFRM', 'ESLT', 'BBVA', 'VRSN', 'ITCI', 'BUD', 'ONC', 'UNM', 'GILD', 'ABBV', 'YUM']

=== Starting Portfolio Optimization ===
Ticker index already set in metrics
Common tickers: ['GILD', 'ABBV']

Calculating covariance matrix...
Calculating expected returns...
Optimizing portfolio...


c:\Users\ping\Files_win10\python\py310\.venv\lib\site-packages\pypfopt\efficient_frontier\efficient_frontier.py:259: UserWarning: max_sharpe transforms the optimization problem so additional objectives may not work as expected.
  warnings.warn(


OptimizationError: ('Please check your objectives/constraints or use a different solver.', 'Solver status: infeasible')

In [ ]:
def select_outperforming_stocks(df):
    """Stock selection with full implementation"""
    # Handle ticker position
    if df.index.name == 'Ticker':
        temp_df = df.reset_index()
    else:
        temp_df = df.copy()

    # Validate required columns
    required_cols = {
        'Ticker', 'Sharpe 30d', 'Sortino 30d', 'Omega 30d',
        'Sharpe 63d', 'Sortino 63d', 'Omega 63d', 'Sharpe 125d',
        'Sortino 125d', 'Omega 125d', 'EPS next Y', 'Profit M',
        'PEG', 'Beta', 'Inst Own', 'Recom'
    }
    missing = required_cols - set(temp_df.columns)
    if missing:
        raise ValueError(f"Missing columns: {missing}")

    # Horizon configuration
    horizons = {
        15: '30d',
        30: '30d', 
        60: '63d',
        120: '125d'
    }

    selected_stocks = {}
    
    for days, suffix in horizons.items():
        # Risk metrics
        sharpe = f'Sharpe {suffix}'
        sortino = f'Sortino {suffix}'
        omega = f'Omega {suffix}'
        
        # Filter stocks
        valid = temp_df[
            (temp_df[sharpe] > 0) &
            (temp_df[sortino] > 0) &
            (temp_df[omega] > 1)
        ].copy()
        
        if valid.empty:
            selected_stocks[f'{days}_days'] = []
            continue

        # Scoring metrics
        scoring = [
            (sharpe, 1.5, False),
            (sortino, 1.5, False),
            (omega, 1.2, False),
            ('EPS next Y', 1.0, False),
            ('Profit M', 0.8, False),
            ('PEG', 0.7, True),
            ('Beta', 0.7, True),
            ('Inst Own', 0.5, False),
            ('Recom', 0.5, True)
        ]

        # Calculate scores
        for metric, weight, invert in scoring:
            z_vals = zscore(valid[metric], nan_policy='omit')
            z_vals = np.nan_to_num(z_vals, nan=0.0)
            if invert:
                z_vals = -z_vals
            valid[f'{metric}_score'] = z_vals * weight

        # Total score
        score_cols = [f'{metric}_score' for metric, _, _ in scoring]
        valid['total_score'] = valid[score_cols].sum(axis=1)
        
        # Select top 5
        top_5 = valid.nlargest(5, 'total_score')['Ticker'].tolist()
        selected_stocks[f'{days}_days'] = top_5[:5]

    return selected_stocks

def optimize_portfolio(selected_tickers, df_metrics, df_correlation):
    """Portfolio optimization with index awareness"""
    # Handle Ticker index in metrics
    if df_metrics.index.name == 'Ticker':
        metrics = df_metrics
    else:
        metrics = df_metrics.set_index('Ticker')
    
    # Common tickers handling
    common = list(set(selected_tickers) 
                  & set(df_correlation.index) 
                  & set(metrics.index))
    
    # Prepare data
    df_corr = df_correlation.loc[common, common]
    df_sub = metrics.loc[common]  # Now using index directly
    
    # Rest of optimization code remains the same...

    """Portfolio optimization with full implementation"""
    # Common tickers
    common = list(set(selected_tickers) & set(df_correlation.index) & set(df_correlation.columns))
    
    # Prepare data
    df_corr = df_correlation.loc[common, common]
    df_sub = df_metrics.set_index('Ticker').loc[common]
    
    # Covariance matrix
    volatilities = df_sub['Volatility M'] * np.sqrt(21)
    cov_matrix = pd.DataFrame(
        np.outer(volatilities, volatilities) * df_corr.values,
        index=common,
        columns=common
    )
    
    # Expected returns
    df_sub['Expected Return'] = (
        0.4 * df_sub['EPS next Y'] +
        0.3 * df_sub['Profit M'] +
        0.2 * df_sub['Sales Q/Q'] +
        0.1 * (5 - df_sub['Recom'])
    )
    
    # Optimize
    ef = EfficientFrontier(
        df_sub['Expected Return'],
        cov_matrix,
        weight_bounds=(0.02, 0.25)
    )
    ef.add_objective(objective_functions.L2_reg, gamma=0.1)
    ef.max_sharpe()
    
    return ef.clean_weights()

def full_analysis_pipeline(df_metrics, df_correlation):
    """Complete pipeline with index safety"""
    # Ensure metrics keeps Ticker index
    if df_metrics.index.name != 'Ticker':
        df_metrics = df_metrics.set_index('Ticker')
    
    # Run selection and optimization
    selected = select_outperforming_stocks(df_metrics)
    all_tickers = [t for sublist in selected.values() for t in sublist]
    
    return optimize_portfolio(all_tickers, df_metrics, df_correlation)

In [32]:
import pandas as pd
import numpy as np
from scipy.stats import zscore
from pypfopt import EfficientFrontier, objective_functions

def select_outperforming_stocks(df):
    """Stock selection with full implementation"""
    # Handle ticker position
    if df.index.name == 'Ticker':
        temp_df = df.reset_index()
    else:
        temp_df = df.copy()

    # Validate required columns
    required_cols = {
        'Ticker', 'Sharpe 30d', 'Sortino 30d', 'Omega 30d',
        'Sharpe 63d', 'Sortino 63d', 'Omega 63d', 'Sharpe 125d',
        'Sortino 125d', 'Omega 125d', 'EPS next Y', 'Profit M',
        'PEG', 'Beta', 'Inst Own', 'Recom'
    }
    missing = required_cols - set(temp_df.columns)
    if missing:
        raise ValueError(f"Missing columns: {missing}")

    # Horizon configuration
    horizons = {
        15: '30d',
        30: '30d', 
        60: '63d',
        120: '125d'
    }

    selected_stocks = {}
    
    for days, suffix in horizons.items():
        # Risk metrics
        sharpe = f'Sharpe {suffix}'
        sortino = f'Sortino {suffix}'
        omega = f'Omega {suffix}'
        
        # Filter stocks
        valid = temp_df[
            (temp_df[sharpe] > 0) &
            (temp_df[sortino] > 0) &
            (temp_df[omega] > 1)
        ].copy()
        
        if valid.empty:
            selected_stocks[f'{days}_days'] = []
            continue

        # Scoring metrics
        scoring = [
            (sharpe, 1.5, False),
            (sortino, 1.5, False),
            (omega, 1.2, False),
            ('EPS next Y', 1.0, False),
            ('Profit M', 0.8, False),
            ('PEG', 0.7, True),
            ('Beta', 0.7, True),
            ('Inst Own', 0.5, False),
            ('Recom', 0.5, True)
        ]

        # Calculate scores
        for metric, weight, invert in scoring:
            z_vals = zscore(valid[metric], nan_policy='omit')
            z_vals = np.nan_to_num(z_vals, nan=0.0)
            if invert:
                z_vals = -z_vals
            valid[f'{metric}_score'] = z_vals * weight

        # Total score
        score_cols = [f'{metric}_score' for metric, _, _ in scoring]
        valid['total_score'] = valid[score_cols].sum(axis=1)
        
        # Select top 5
        top_5 = valid.nlargest(5, 'total_score')['Ticker'].tolist()
        selected_stocks[f'{days}_days'] = top_5[:5]

    return selected_stocks

def optimize_portfolio(selected_tickers, df_metrics, df_correlation):
    """Portfolio optimization with full implementation"""
    # Common tickers
    common = list(set(selected_tickers) & set(df_correlation.index) & set(df_correlation.columns))
    
    # Prepare data
    df_corr = df_correlation.loc[common, common]
    df_sub = df_metrics.set_index('Ticker').loc[common]
    
    # Covariance matrix
    volatilities = df_sub['Volatility M'] * np.sqrt(21)
    cov_matrix = pd.DataFrame(
        np.outer(volatilities, volatilities) * df_corr.values,
        index=common,
        columns=common
    )
    
    # Expected returns
    df_sub['Expected Return'] = (
        0.4 * df_sub['EPS next Y'] +
        0.3 * df_sub['Profit M'] +
        0.2 * df_sub['Sales Q/Q'] +
        0.1 * (5 - df_sub['Recom'])
    )
    
    # Optimize
    ef = EfficientFrontier(
        df_sub['Expected Return'],
        cov_matrix,
        weight_bounds=(0.02, 0.25)
    )
    ef.add_objective(objective_functions.L2_reg, gamma=0.1)
    ef.max_sharpe()
    
    return ef.clean_weights()

def full_analysis_pipeline(df_metrics, df_correlation):
    """Complete working pipeline"""
    # Step 1: Stock selection
    selected = select_outperforming_stocks(df_metrics)
    all_tickers = list(set(
        t for horizon in selected.values() 
        for t in horizon if isinstance(t, str)
    ))
    
    # Step 2: Portfolio optimization
    if not all_tickers:
        raise ValueError("No stocks selected in initial filtering")
        
    return optimize_portfolio(all_tickers, df_metrics, df_correlation)

# Usage:
# df_metrics = pd.read_pickle(r'..\data\df_finviz_stocks_n_ratios.pkl')
# df_correlation = pd.read_csv('correlation_matrix.csv', index_col=0)
# portfolio = full_analysis_pipeline(df_metrics, df_correlation)
# print(portfolio)

In [ ]:
import pandas as pd
import numpy as np
from scipy.stats import zscore

def select_outperforming_stocks(df):
    """
    Analyze stock metrics to select top 5 outperforming stocks with acceptable risk
    across multiple time horizons. Handles Tickers both as columns and index.
    """
    # Validate input and handle ticker position
    if not isinstance(df, pd.DataFrame):
        raise TypeError("Input must be a pandas DataFrame")
    
    # Create working copy and check ticker position
    temp_df = df.copy()
    if 'Ticker' not in temp_df.columns:
        if temp_df.index.name == 'Ticker':
            temp_df = temp_df.reset_index()
        else:
            raise ValueError("Ticker information not found in columns or index")

    # Check required columns
    required_cols = {
        'Ticker', 'Sharpe 30d', 'Sortino 30d', 'Omega 30d',
        'Sharpe 63d', 'Sortino 63d', 'Omega 63d', 'Sharpe 125d',
        'Sortino 125d', 'Omega 125d', 'EPS next Y', 'Profit M',
        'PEG', 'Beta', 'Inst Own', 'Recom'
    }
    
    missing = required_cols - set(temp_df.columns)
    if missing:
        raise ValueError(f"Missing required columns: {missing}")

    # Horizon configuration
    horizons = {
        15: ('30d', 'short'),
        30: ('30d', 'short'),
        60: ('63d', 'medium'),
        120: ('125d', 'long')
    }

    results = {}
    
    for days, (suffix, _) in horizons.items():
        sharpe = f'Sharpe {suffix}'
        sortino = f'Sortino {suffix}'
        omega = f'Omega {suffix}'
        
        # Filter stocks with positive risk-adjusted returns
        valid_stocks = temp_df[
            (temp_df[sharpe] > 0) &
            (temp_df[sortino] > 0) &
            (temp_df[omega] > 1)
        ].copy()
        
        if valid_stocks.empty:
            results[f'{days}_days'] = []
            continue

        # Create scoring system
        scoring_metrics = [
            (sharpe, 1.5, False),
            (sortino, 1.5, False),
            (omega, 1.2, False),
            ('EPS next Y', 1.0, False),
            ('Profit M', 0.8, False),
            ('PEG', 0.7, True),
            ('Beta', 0.7, True),
            ('Inst Own', 0.5, False),
            ('Recom', 0.5, True)
        ]

        # Calculate weighted z-scores
        for metric, weight, invert in scoring_metrics:
            z_values = zscore(valid_stocks[metric], nan_policy='omit')
            z_values = np.nan_to_num(z_values, nan=0.0)
            if invert:
                z_values = -z_values
            valid_stocks[f'{metric}_score'] = z_values * weight

        # Calculate total score
        score_cols = [f'{metric}_score' for metric, _, _ in scoring_metrics]
        valid_stocks['total_score'] = valid_stocks[score_cols].sum(axis=1)
        
        # Select top 5
        top_5 = valid_stocks.nlargest(5, 'total_score')['Ticker'].tolist()
        results[f'{days}_days'] = top_5[:5]

    return results

In [23]:
recommendations = select_outperforming_stocks(df)

In [24]:
recommendations

{'15_days': ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN'],
 '30_days': ['BUD', 'ITCI', 'ABBV', 'YUM', 'VRSN'],
 '60_days': ['ITCI', 'ESLT', 'VRSN', 'BBVA', 'ONC'],
 '120_days': ['ESLT', 'AFRM', 'ITCI', 'GILD', 'UNM']}